### PR TITLE
Cleanup + Optimize Dockerfile

### PR DESCRIPTION
**Pull request by snickell**
Refactor of `docker/dockerfiles/Dockerfile` and its entrypoint.sh.

- Consolidate `apt-get` commands
- Use multi-stage build to parallelize slower operations: node compile and ruby compile. Multi-stage builds also mean unrelated changes don't trigger cache-misses and cause rebuilds (less rebuilding of node & ruby = people more freely do the file they want rather than adding on commits to the end to avoid waiting an extra hour).
- Parameterize $USER and $GROUP rather than hardcoding old "circleci" values (still using circleci, but now its an env variable change docker-wise)
- Use a consistent formatting style throughout (selected from the several in play before)
- Smaller final image, fewer layers (faster image pulls => slightly faster drone builds)

Extracted from https://github.com/code-dot-org/code-dot-org/pull/51533 .

This PR contains no version updates to .drone.yml etc: as this PR doesn't add any features or advantages (except for development!), the intention is to deploy both of these at the same time to avoid "no extra features" deploy churn.

### CODE REVIEW

File: docker/dockerfiles/Dockerfile
Comment (Hamms) [inline docker/dockerfiles/Dockerfile:8]: has this functionality been removed? I'm not seeing an equivalent in the new version

File: docker/dockerfiles/Dockerfile
Comment (Hamms) [inline docker/dockerfiles/Dockerfile]: Should this be

```suggestion
# compile node to /build, this layer is merged at the end
```

File: docker/dockerfiles/Dockerfile
Comment (Hamms) [inline docker/dockerfiles/Dockerfile:135]: are the `DEBIAN_FRONTEND=noninteractive` directives no longer necessary?

File: docker/dockerfiles/entrypoint.sh
Comment (Hamms) [inline docker/dockerfiles/entrypoint.sh]: How does [invoking tests like `/entrypoint.sh docker/unit_tests.sh || EXIT_CODE=$?`]( https://github.com/code-dot-org/code-dot-org/pull/51630/files#diff-b54b39f1afced2465e1f3641db9d5bbf4f3a7fcf890996dfedd3c197bcb7f8c7R70) work without this `exec` line?

**Review by Hamms — COMMENTED** [review]
Couple of nits and questions I'd like to get addressed before full approval, but generally this looks great! Very nice to see this dockerfile getting some much-needed attention, and as a bonus I learned a lot about multi-stage builds.

File: docker/dockerfiles/Dockerfile
Comment (snickell) [reply docker/dockerfiles/Dockerfile:135]: I set them as an ENV at the top, so they apply to all apt-get commands rather than being remembered in some and forgotten in others. If you miss one, all the debconf that's backlogged can drop on your head, so its better to go all-in on noninteractive to prevent future surprises.

```
ENV \
    LANG=C.UTF-8 \
    TZ=Etc/UTC \
    DEBIAN_FRONTEND=noninteractive \
    USER=circleci \
    GROUP=circleci
```

File: docker/dockerfiles/Dockerfile
Comment (snickell) [reply docker/dockerfiles/Dockerfile:8]: This is handled using the TZ env var at the top instead, which is prettier and "more docker" in the sense that it allows you to vary the TZ for a container at start (not something we want to do on drone, but makes the dockerfile more re-usable for e.g. dev):

```
ENV \
    LANG=C.UTF-8 \
    TZ=Etc/UTC \
```

File: docker/dockerfiles/entrypoint.sh
Comment (snickell) [reply docker/dockerfiles/entrypoint.sh]: gah, I mismerged somewhere here, TY for catching 🙏🙏🙏

File: docker/dockerfiles/Dockerfile
Comment (Hamms) [reply docker/dockerfiles/Dockerfile:135]: Oh, very nice! Yeah, much prefer this being set globally. Thanks!

File: .drone.yml
Comment (snickell) [inline .drone.yml:148]: This makes `ui-tests` behavior consistent with `unit-tests` (which already is set to pull: always), so they're both running on the same version of the `codedotorg/code-dot-org` image if/when there are multiple pushes to the same tag.

**Review by Hamms — COMMENTED** [review]
Looks like drone still isn't running any tests on this branch, even after restoration of the `exec` line:

```
1.6.0: Pulling from codedotorg/code-dot-org
Digest: sha256:2688fead9a062dfb3c3540bc19106f462df4c2de480d351e068f0506c335d014
Status: Image is up to date for codedotorg/code-dot-org:1.6.0
+ sudo chown -R circleci:circleci .
+ cp -rn "$(pwd)/home/circleci/.rbenv" /home/circleci || true
+ rm -rf "$(pwd)/home/circleci/.rbenv"
+ export CIRCLE_SHA1=$DRONE_COMMIT
+ EXIT_CODE=0
+ /entrypoint.sh docker/unit_tests.sh || EXIT_CODE=$?
/entrypoint.sh: 3: cd: can't cd to /home/circleci/code-dot-org
fixuid: fixuid should only ever be used on development systems. DO NOT USE IN PRODUCTION
fixuid: runtime UID '3434' already matches container user 'circleci' UID
fixuid: runtime GID '3434' already matches container group 'circleci' GID
chown: cannot access '/home/circleci/.config': No such file or directory
chown: cannot access '/home/circleci/.cache': No such file or directory
+ while pgrep bash >/dev/null; do echo "found bash process; keeping container alive"; sleep 20; done
+ exit $EXIT_CODE
```

**Comment by snickell** [comment]
> Looks like drone still isn't running any tests on this branch, even after restoration of the `exec` line

EBKAC: missing a `docker push` after restoring `entrypoint.sh`. 

Drone tests are finally kicking off for reals, TY for 🦅  eyes thru rinse/review cycles 🙇

**Comment by snickell** [comment]
Gah, because no JS/ruby files were changed, it skipped all the tests 🤦 To trigger a full test run, I made a trivial change to `lib/Rakefile` and `apps/README.md` thus bypassing `run_tests_if_changed` caching test results when file patterns don't change.

Results of the cache-busted drone unit-tests will be here: https://drone.cdn-code.org/code-dot-org/code-dot-org/28612

Now reverting the cache-busting changes:

**Comment by snickell** [comment]
Tests did not pass, looking into this tomorrow:
```
  test_fork                                                      ERROR (0.00s)
Minitest::UnexpectedError:         NotImplementedError: fork() function is unimplemented on this machine
            /drone/src/lib/test/cdo/test_buffer.rb:72:in `fork'
            /drone/src/lib/test/cdo/test_buffer.rb:72:in `block in test_fork'
```

Weird:
```
circleci@c953a25e57a3:~$ ruby --version
ruby 2.7.5p203 (2021-11-24 revision f69aeb8314) [x86_64-linux]
circleci@c953a25e57a3:~$ irb
irb(main):001:0> fork
Traceback (most recent call last):
        5: from /home/circleci/.rbenv/versions/2.7.5/bin/irb:23:in `<main>'
        4: from /home/circleci/.rbenv/versions/2.7.5/bin/irb:23:in `load'
        3: from /home/circleci/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        2: from (irb):1
        1: from (irb):1:in `fork'
NotImplementedError (fork() function is unimplemented on this machine)
```

**Comment by snickell** [comment]
This is an actual meaty behavior difference between built docker images:

code-dot-org:1.5.4:
```
➜ docker run -it codedotorg/code-dot-org:1.5.4 bash -c "rbenv global 2.7.5 && ruby -e 'fork' && echo 'fork worked!'"
fork worked!
```

code-dot-org:1.6.0 (this PR):
```
➜  docker run -it codedotorg/code-dot-org:1.6.0 bash -c "rbenv global 2.7.5 && ruby -e 'fork' && echo 'fork worked!'"
Traceback (most recent call last):
        1: from -e:1:in `<main>'
-e:1:in `fork': fork() function is unimplemented on this machine (NotImplementedError)
```

However, after a rebuild of ruby 2.7.5 (still using rbenv) on the 1.6 image, fork DOES work. Aie. What happened???

**Comment by snickell** [comment]
deb `libaio` has to be installed before building ruby, or it builds a ruby w/o a working `fork` (!!!)

**Comment by snickell** [comment]
Real tests pass (45 minutes for unit), available here: https://drone.cdn-code.org/code-dot-org/code-dot-org/28617

Reverted the cache busting changes.

**Review by Hamms — APPROVED** [review]
Lol, what a ride. I had never considered that we might end up with a ruby that installed successfully but without some aspects of the language enabled. Nice catch!

This change LGTM! :shipit:

**Pull request merged**
